### PR TITLE
File caching

### LINKED ISSUE #43: Caching of files
When opening a file a second time, then the download starts again (for some reason it starts at 50%), but that's still wasted time and bandwith.

1. If a file was not modified (hash is not modified) and it's still saved locally, then open it directly
  - variant A: calculate hash after clicking the file to check if it can be reused
  - variant B: save files locally with random filenames and keep them at a database, as the user uses syncthing lite to access it, he does not even see this; moreover, this would make it easier to limit the cache size as the database could track file sizes and their last usage
2. If the user does not need the new file version, add an cancel option to the download dialog/ add an "use local version" to the context menu
  - variant B would fit better here, it allows downloading to a new file while keeping the old one intact
3. When the user does not have an connection to any PC who provides the file, show a warning and open the cached copy after that/ show an error message saying that no PC providing this file could be reached

### PR DESCRIPTION
**Pull request by l-jonas**
This depends on https://github.com/syncthing/syncthing-lite/pull/57 (but can be rebased)

The files were already cached (saved locally), but it was not tried to use them. This names files by their hash to ensure that no old file version is used. The side effect of this is that equal files are only downloaded and saved once.

This closes #43

What's missing:

- delete local files which are no longer provided by the server
- delete local files which were not used for a long time
- delete all local files when the menu item to clear the cache is used

### CODE REVIEW

File: syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt [hunk 1]
```diff
@@ -14,145 +15,175 @@
...
+            }) {
+                override fun close() {
+                    super.close()
+
+                    // delete all temp blocks now
+                    // they are deleted after reading, but the consumer could stop before reading the whole stream
+                    tempRepository.deleteTempData(blockTempIdByHash.values.toList())
+                }
             }
+        } catch (ex: Exception) {
```
Comment (Nutomic) [inline syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: You shouldn't catch general exceptions, that might hide things like a NullPointerException.

Comment (Nutomic) [reply syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt]: Ah I see. I also missed that you are re-throwing the exception, so this is fine.

**Comment by Nutomic** [comment]
Great to have this, thanks :)

Do you think we're ready for a new release of the app now?

**Comment by l-jonas** [comment]
@Nutomic Yes (but it should be still a "0.x.x" version). To reduce bad ratings, I added a note about the reconnect issue (https://github.com/syncthing/syncthing-lite/pull/60), which should be included in a release.

**Comment by Nutomic** [comment]
Agreed, I would probably go for 0.3. Seeing as you know the project best, feel free to make the release yourself. Just update versionCode and versionName, tag the commit and write the changelog. F-Droid should pick the release up automatically, but I'm not sure about Google Play. 

ping @AudriusButkevicius How can we release a new app version to Google Play?

**Comment by AudriusButkevicius** [comment]
You should still have access to the play console. You can re-list the app and upload new apks. Or your could try and setup CI on teamcity to do that for you.

**Comment by AudriusButkevicius** [comment]
Wrapper builds now happen in teamcity.

**Pull request merged**